csharp/ql/src/semmle/code/csharp/dataflow/SignAnalysis.qll

```diff
@@ -0,0 +1,595 @@
+  }
+
+  Sign rem(Sign s) {
+    result = TZero() and s = TNeg()
+    or
+    result = TZero() and s = TPos()
+    or
+    result = this and s = TNeg()
+    or
+    result = this and s = TPos()
+  }
+
+  Sign bitand(Sign s) {
+    result = TZero() and this = TZero()
+    or
+    result = TZero() and s = TZero()
+    or
+    result = TZero() and this = TPos()
+    or
+    result = TZero() and s = TPos()
+    or
+    result = TNeg() and this = TNeg() and s = TNeg()
+    or
+    result = TPos() and this = TNeg() and s = TPos()
+    or
+    result = TPos() and this = TPos() and s = TNeg()
+    or
+    result = TPos() and this = TPos() and s = TPos()
+  }
+
+  Sign bitor(Sign s) {
+    result = TZero() and this = TZero() and s = TZero()
+    or
+    result = TNeg() and this = TNeg()
+    or
+    result = TNeg() and s = TNeg()
+    or
+    result = TPos() and this = TPos() and s = TZero()
+    or
+    result = TPos() and this = TZero() and s = TPos()
+    or
+    result = TPos() and this = TPos() and s = TPos()
+  }
+
+  Sign bitxor(Sign s) {
+    result = TZero() and this = s
+    or
+    result = this and s = TZero()
+    or
+    result = s and this = TZero()
+    or
+    result = TPos() and this = TPos() and s = TPos()
+    or
+    result = TNeg() and this = TNeg() and s = TPos()
+    or
+    result = TNeg() and this = TPos() and s = TNeg()
+    or
+    result = TPos() and this = TNeg() and s = TNeg()
+  }
+
+  Sign lshift(Sign s) {
+    result = TZero() and this = TZero()
+    or
+    result = this and s = TZero()
+    or
+    this != TZero() and s != TZero()
+  }
+
+  Sign rshift(Sign s) {
+    result = TZero() and this = TZero()
+    or
+    result = this and s = TZero()
+    or
+    result = TNeg() and this = TNeg()
+    or
+    result != TNeg() and this = TPos() and s != TZero()
+  }
+}
+
+/** Gets the sign of `e` if this can be directly determined. */
+private Sign certainExprSign(Expr e) {
+  exists(int i | e.(ConstantIntegerExpr).getIntValue() = i |
+    i < 0 and result = TNeg()
+    or
+    i = 0 and result = TZero()
+    or
+    i > 0 and result = TPos()
+  )
+  or
+  not exists(e.(ConstantIntegerExpr).getIntValue()) and
+  (
+    exists(float f |
+      f = e.(LongLiteral).getValue().toFloat() or
+      f = e.(RealLiteral).getValue().toFloat()
+    |
+      f < 0 and result = TNeg()
+      or
+      f = 0 and result = TZero()
+      or
+      f > 0 and result = TPos()
+    )
+    or
+    exists(string charlit | charlit = e.(CharLiteral).getValue() |
+      if charlit = "\\0" or charlit = "\\u0000" then result = TZero() else result = TPos()
+    )
+    or
+    containerSizeAccess(e.(PropertyAccess)) and
+    (result = TPos() or result = TZero())
+    or
+    e instanceof Linq::CountCall and
+    (result = TPos() or result = TZero())
+  )
+}
+
+private predicate containerSizeAccess(PropertyAccess pa) {
+  propertyOverrides(pa.getTarget(), "System.Collections.Generic.ICollection<>", "Count") or
```

Comment:
This should probably go to a helper file, because we use it in `ContainerSizeCmpZero.ql` too. But I couldn't find a good location for it. (Also, the check is using a slightly different variant.